File: src/lib/types/gacha.ts
export enum GachaType {
    CHARACTER,
    WEAPON
}


export enum GachaMovieType {
    NORMAL,
    GUARANTEE
}


export interface GachaPoolItem {
    id: number,
    rank: number,
    odds: number,
    isRateUp: boolean,
    isLimited?: boolean,
    isExchangeable?: boolean,
    trialReadingForced?: boolean,
    rarity: number
}


export interface GachaRankRates {
    normal: number[],
    multiGuarantee: number[]
}


export interface Gacha {
    type: GachaType,
    paymentType: number,
    pageKind?: number,
    singleCost: number,
    multiCost: number,
    discountCost: number,
    tenTimesPerAccountCost?: number,
    onceTicketItemId?: number,
    tenTicketItemId?: number,
    wildcardTicketAvailable?: boolean,
    rarityOddsId?: string,
    guaranteeRarity?: number,
    rankRates?: GachaRankRates,
    equipmentMovieProbabilityId?: string,
    startDate: string,
    endDate: string,
    pool: Record<string, GachaPoolItem[]>
}


export interface CharacterGacha extends Gacha {
    movieName: string,
    guaranteeMovieName: string,
    toUseOddsUpAsTrialReading?: boolean,
    canBeStartDashExchange?: boolean
}


export type Gachas = Record<string, Gacha>


export type GachaDrawResult = number[]


export interface RewardPlayerGachaDrawResult {
    draw: GachaDraws,
    characters: Object[],
    equipment: Object[],
    items: Record<number, number>,
    isErupt?: boolean
}


export interface GachaCharacterDraw {
    character_id: number,
    movie_id: string,
    seed: number,
    entry_count: number,
    ex_boost_item?: {
        id: number,
        count: number
    } | []
}


export interface GachaEquipmentDraw {
    equipment_id: number,
    treasure_up_type: number
}


export type GachaDraws = (GachaCharacterDraw | GachaEquipmentDraw)[]



// shops

File: scripts/converters/gacha/gacha.py
import os
import util
from .rarities import convert_gacha_rarities

def convert_gacha(obj):
    converted = {}
    for gacha_id, gacha_data in obj.items():
        payment_type = int(gacha_data[4])

        # 0 = character; 1 = weapon
        gacha_type = int(gacha_data[13])
        if payment_type == 0:
            single_cost = int(gacha_data[5])
            multi_cost = int(gacha_data[6])
            discount_single_cost = int(gacha_data[7])

            if gacha_type == 0:
                converted_gacha = {
                    "type": gacha_type,
                    "paymentType": payment_type,
                    "singleCost": single_cost,
                    "multiCost": multi_cost,
                    "discountCost": discount_single_cost,
                    "movieName": gacha_data[17],
                    "guaranteeMovieName": gacha_data[18],
                    "startDate": gacha_data[29],
                    "endDate": gacha_data[30]
                }
                converted_gacha["pool"] = {
                    "3": convert_gacha_rarities(os.path.join(util.FILE_INPUT, "gacha_odds", f"{gacha_data[14]}.json")),
                    "2": convert_gacha_rarities(os.path.join(util.FILE_INPUT, "gacha_odds", f"{gacha_data[15]}.json")),
                    "1": convert_gacha_rarities(os.path.join(util.FILE_INPUT, "gacha_odds", f"{gacha_data[16]}.json")),
                }
                converted[gacha_id] = converted_gacha 

            elif gacha_type == 1:
                converted_gacha = {
                    "type": gacha_type,
                    "paymentType": payment_type,
                    "singleCost": single_cost,
                    "multiCost": multi_cost,
                    "discountCost": discount_single_cost,
                    "startDate": gacha_data[29],
                    "endDate": gacha_data[30]
                }
                converted_gacha["pool"] = {
                    "3": convert_gacha_rarities(os.path.join(util.FILE_INPUT, "gacha_odds", f"{gacha_data[22]}.json")),
                    "2": convert_gacha_rarities(os.path.join(util.FILE_INPUT, "gacha_odds", f"{gacha_data[23]}.json")),
                    "1": convert_gacha_rarities(os.path.join(util.FILE_INPUT, "gacha_odds", f"{gacha_data[24]}.json")),
                }
                converted[gacha_id] = converted_gacha
    
    return converted
